tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/trex/TransitionExecutorImpl.scala

```diff
@@ -0,0 +1,388 @@
+  }
+
+  /**
+    * Pick non-deterministically one transition among the transitions that are prepared
+    * in the current step. Further, assume that the picked transition has fired.
+    * This method must be called after at least one call to prepareTransition.
+    */
+  override def pickTransition(): Oracle = {
+    assert(controlState == Preparing())
+    // assert that there is at least one prepared transition
+    logger.info("Step %d, level %d: picking a transition out of %d transition(s)"
+      .format(stepNo, ctx.rewriter.contextLevel, preparedTransitions.size))
+    assert(preparedTransitions.nonEmpty)
+    val sortedTransitions = preparedTransitions.toSeq.sortBy(_._1)
+
+    // pick an index j \in { 0..k } of the fired transition
+    val picker = new CherryPick(ctx.rewriter)
+    val (oracleState, oracle) = picker.oracleFactory.newDefaultOracle(lastState, sortedTransitions.length)
+
+    if (sortedTransitions.isEmpty) {
+      throw new IllegalArgumentException("unable to pick transitions from empty set")
+    } else if (sortedTransitions.lengthCompare(1) == 0) {
+      ctx.solver.assertGroundExpr(sortedTransitions.head._2.trigger.toNameEx)
+      lastState = oracleState.setBinding(sortedTransitions.head._2.binding)
+      val transitionNo = sortedTransitions.head._1
+      // use a fixed transition
+      val mockOracle = new MockOracle(transitionNo)
+      revStack = (lastState.binding.shiftBinding(consts), mockOracle) :: revStack
+    } else {
+      // if oracle = i, then the ith transition is enabled
+      ctx.solver.assertGroundExpr(oracle.caseAssertions(oracleState, sortedTransitions.map(_._2.trigger.toNameEx)))
+
+      // glue the computed states S_0, ..., S_k together:
+      // for every variable x', pick c_x from { S_1[x'], ..., S_k[x'] }
+      //   and require \A i \in { 0.. k-1}. j = i => c_x = S_i[x']
+      // Then, the final controlState binds x' -> c_x for every x' \in Vars'
+      var nextState = oracleState
+
+      def pickVar(x: String): ArenaCell = {
+        val toPickFrom = sortedTransitions map (p => p._2.binding(x))
+        nextState = picker.pickByOracle(nextState,
+          oracle, toPickFrom, nextState.arena.cellFalse().toNameEx) // no else case
+        nextState.asCell
+      }
+
+      def getAssignedVars(binding: Binding) = binding.forgetNonPrimed(Set()).toMap.keySet
+      val primedVars = getAssignedVars(sortedTransitions.head._2.binding) // only VARIABLES, not CONSTANTS
+
+      val finalVarBinding = Binding(primedVars.toSeq map (n => (n, pickVar(n))): _*) // variables only
+      val constBinding = Binding(oracleState.binding.toMap.filter(p => consts.contains(p._1)))
+      lastState = nextState.setBinding(finalVarBinding ++ constBinding)
+      // the sparse oracle is mapping the oracle values to the transition numbers
+      val sparseOracle = new SparseOracle(oracle, preparedTransitions.keySet)
+      revStack = (lastState.binding.shiftBinding(consts), sparseOracle) :: revStack
```

Comment:
I guess the same nits re: naming and abstracting pushing to the stack could apply through this file, but that's non-essential of course.

Reply:
Good point. Fixed that too.